pkg/stores/proxy/proxy_store.go

```diff
@@ -370,6 +370,7 @@ func (s *Store) WatchNames(apiOp *types.APIRequest, schema *types.APISchema, w t
 				} else {
 					logrus.Debugf("WatchNames received error: %v", item)
 				}
+				result <- item
```

Comment:
I'm a bit concerned on the security of this option. From the [go docs](https://pkg.go.dev/k8s.io/apimachinery@v0.29.0/pkg/watch#Event) for the `Object` field of the `watch.Event` type:

```
//  * If Type is Error: *api.Status is recommended; other types may make sense
//    depending on context.
```

This to me indicates that this could contain a real object. 

AFAIK, The function that this is using (`WatchNames`) is used when the user doesn't have access to all objects, but still needs to watch for the resources that they do have access to. To achieve this, this function [filters](https://github.com/rancher/steve/blob/9935404019063126b6ac856db1ad56e0bc457ff8/pkg/stores/proxy/proxy_store.go#L376) the objects by names to ensure that the user has access to the object in the channel. The user only gets the object back if it has access to that object. However, since this filtering happens after the error handling logic, this new code could theoretically return an object that the user doesn't have access to.

Reply:
You are right that sending arbitrary objects down the `result` channel in the `Object` field of `watch.Event` is not prevented by types alone.

Nevertheless I believe the use is safe here, because:

- `item` is only received from the `c` channel, which is created and returned by [watch](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L358)
- in turn, `watch` only passes `c` to [listAndWatch](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L405) after its creation and before its closing. Therefore, `listAndWatch` is the only starting place where values can be sent to `c`
- `listAndWatch` only sends values in 5 places:
  - [three](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L296) [of](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L317) [them](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L328) via the `returnErr` function, which [crafts a new watch.Event](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L264C12-L269) with `Type == watch.Error` with an `api.Status` as its `Object`. Therefore these objects are safe to return
  - [two](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L315) of [them](https://github.com/rancher/steve/blob/2365164708b72f5c10dd5a38799543a4c28aa793/pkg/stores/proxy/proxy_store.go#L337) via direct send. In both cases code checks immediately before that is not an error. Therefore in this case the code added by this PR is not executed


So that's my proof by exhaustion. I understand something other than a proof by exhaustion might be more desirable, but I'd argue such a bigger refactoring should be separate from this PR, which is meant to include the bugfix exclusively.

Reply:
Thanks for giving that a look. I think that I agree that we don't need function-level safety here, but I'm a bit concerned about one additional case that's not covered here. [listAndWatch](https://github.com/rancher/steve/blob/9935404019063126b6ac856db1ad56e0bc457ff8/pkg/stores/proxy/proxy_store.go#L329-L333) currently has a case that looks like it functions much the same as the case that you fixed does (i.e. debug the error and don't send on the channel). Would you mind seeing if you need to fix that as well? If so, how you handle that case will determine if this is safe since this will be another place where a value is sent on the channel,.

Reply:
It is indeed a similar. I'll open a follow-up PR to keep that case separate, as the user case I am working for isn't experiencing that particular error (I can tell from the debug logs not containing the string logged there).

Reply:
I'll resolve this thread for now - if you can look at the other part relatively soon I would appreciate it, so that we don't forget about the issue here.